### PR TITLE
Visual feedback around tar/zsync on Kobo during OTA updates

### PR DESCRIPTION
**Pull request by NiLuJe**
Requires a base bump to pickup FBInk, I'm waiting for #4040 to land first ;).

Minor fix to the version normalizer pattern so that it behaves properly with both older and current version strings (it was picking up the date suffix in the commit, and did not expect there to be >999 revisions since the last tag ^^).

### CODE REVIEW

**Comment by NiLuJe** [comment]
Turns out they do reply w/ a 206 (but it's still borked anyway) when zsync does its thing... which is why it did not abort.

**Comment by NiLuJe** [comment]
The good news is: now I know why I never could get OTA updates to work right... I was always trying to use one of my borked mirror :D.

**Comment by cramoisi** [comment]
Would be nice !

File: platform/kobo/koreader.sh [hunk 1]
```diff
@@ -11,9 +11,19 @@ cd "${KOREADER_DIR}" || exit
...
-    # TODO: any graphic indication for the updating progress?
-    ./tar xf "${NEWUPDATE}" --strip-components=1 --no-same-permissions --no-same-owner \
-        && mv "${NEWUPDATE}" "${INSTALLED}"
+    # shellcheck disable=SC2016
+    ./tar xf "${NEWUPDATE}" --strip-components=1 --no-same-permissions --no-same-owner --checkpoint=200 --checkpoint-action=exec='./kotar_cpoint $TAR_CHECKPOINT'
+    fail=$?
+    # Cleanup behind us...
+    if [ "${fail}" -eq 0 ]; then
+        mv "${NEWUPDATE}" "${INSTALLED}"
+        ./fbink -q -y -6 -pm "Update successful :)"
```
Comment (Frenzie) [inline platform/kobo/koreader.sh]: lol, cute :P

**Comment by NiLuJe** [comment]
I should have rebased earlier to pickup the original readerlink test fix, oops :D.

**Comment by cramoisi** [comment]
@NiLuJe (Do you think you could make the ota update working if there is some (plugin) directories missing on the device ?) :D

**Comment by NiLuJe** [comment]
@cramoisi : I remember discussing something like this with you a few weeks ago, what were your exact symptoms? because all my trouble basically went away once I figured out the culprit was my OpenStack mirrors :/.

**Comment by cramoisi** [comment]
@NiLuJe : I removed most of the plugin I don't use from the plugins directory. When OTA update, the check fail because the directories are not present.

**Comment by poire-z** [comment]
Another candidate for a manual setting?
I have this simple thing in my own patch:
```lua
--- orig_koreader/frontend/pluginloader.lua     2018-06-15 08:01:28.000000000 +0200
+++ koreader/frontend/pluginloader.lua  2018-06-15 09:08:08.000000000 +0200
@@ -27,6 +27,25 @@
 function PluginLoader:loadPlugins()
     if self.plugins then return self.plugins end

+    -- on ignore certains plugins
+    local plugindir_to_ignore = {}
+    plugindir_to_ignore["autofrontlight.koplugin"] = true
+    plugindir_to_ignore["backgroundrunner.koplugin"] = true
+    plugindir_to_ignore["calibrecompanion.koplugin"] = true
[... a bit ashamed to show how many I ignore ...]
+
     self.plugins = {}
     local lookup_path_list = { DEFAULT_PLUGIN_PATH }
     local extra_paths = G_reader_settings:readSetting("extra_plugin_paths")
@@ -56,7 +75,7 @@
             local plugin_root = lookup_path.."/"..entry
             local mode = lfs.attributes(plugin_root, "mode")
             -- valid koreader plugin directory
-            if mode == "directory" and entry:find(".+%.koplugin$") then
+            if mode == "directory" and entry:find(".+%.koplugin$") and not plugindir_to_ignore[entry] then
                 local mainfile = plugin_root.."/main.lua"
                 package.path = string.format("%s/?.lua;%s", plugin_root, package_path)
                 package.cpath = string.format("%s/lib/?.so;%s", plugin_root, package_cpath)
```
Easy to adapt to use from settings.reader.lua:
```
    ["plugins_disabled"] = {
        ["autofrontlight"] = true,
        ["backgroundrunner"] = true,
        ["calibrecompanion"] = true
    },
```

(There may be dependancies between plugins, so that requires the person to know a bit about them, so good to stay a manual setting.)

**Comment by cramoisi** [comment]
cool ! background-runner shouldn't be remove it's the only one I think.

**Comment by NiLuJe** [comment]
Right, the other solution would be patching the index file too, which is decidedly uglier ;).

**Comment by poire-z** [comment]
Done in #4054 (for use until we have a plugin management UI :)

**Comment by NiLuJe** [comment]
I was somewhat expecting the landscape crap on some Kobos... But the yellow thing (as well as the effect it had on progress bar there) is extra weird...

Can you try without the `-q` flag and post the console output?

And, if possible, can you try the same from within nickel after a reboot?

**Comment by poire-z** [comment]
Oh, that yellow thing may be me (I had enabled a setting "Disable uneeded RGB>BGR conversion" :| ).
It displays as cyan now with that disabled:
<kbd>reader_2018-jul-05_104156</kbd>

```
# ./fbink -y -6 -pm "Update successful :)" blah bluh
[FBInk] Variable fb info: 1448x1072, 16bpp @ rotation: 0
[FBInk] Fontsize set to 32x32.
[FBInk] Line length: 45 cols, Page size: 33 rows.
[FBInk] Fixed fb info: smem_len 6782976, line_length 2944
[FBInk] Kernel's HZ value appears to be 100
Printing string 'Update successful :)' @ column 0, row -6 (inverted: false, flashing: false, centered: true, left padded: true, clear screen: false, font scaling: x0)
Printing string 'blah' @ column 0, row -5 (inverted: false, flashing: false, centered: true, left padded: true, clear screen: false, font scaling: x0)
Printing string 'bluh' @ column 0, row -4 (inverted: false, flashing: false, centered: true, left padded: true, clear screen: false, font scaling: x0)
```

> And, if possible, can you try the same from within nickel after a reboot?

My wifi doesn't stick when I switch to nickel (can't set it up in nickel, static IP, etc...). I'll have to launch a bg script to do that, and I'm a bit late for work :| This evening if still needed

**Comment by NiLuJe** [comment]
Nope, yellow was right ^^.

I can somewhat replicate this by interrupting on-animator at the right time (because pickel apparently messes with the framebuffer in two different ways).... with the added quirk that rotation is broken slightly differently :D.

**Comment by NiLuJe** [comment]
So I can at least try to fix the color issue, but I don't think I'll do anything for rotation, because Kobo is the outlier here (everybody else does it right, including Kobo themselves in later FW ;)).

Assuming that if kobo, bpp == 16, xres > yres, and rotate = 0 means it's a broken landscape that needs to be corrected *might* work, but it's fragile at best (i.e., I've seen my H2O boot in CCW instead of CW, for instance, and on these broken FW, since rotate lies, there's no way to detect that).

**Comment by NiLuJe** [comment]
Yeah, color was a genuinely stupid bug (data type truncation).

**Comment by NiLuJe** [comment]
At least the smiley's cute that way? :D.

**Comment by NiLuJe** [comment]
When you have the chance, it'd be interesting to know if Nickel indeed does anything with the rotation, but, yeah, not gonna touch that with a ten foot pole in FBInk, fixing the colors was mainly what I was after for those devices ;).

**Comment by poire-z** [comment]
I tried, but it seems fbink will wait till way after I exit nickel to display its text over KSM screen.
If I launch this script:
```sh
#!/bin/sh
(
date
echo Sleeping
sleep 60
echo
date
echo Launching
/mnt/onboard/.adds/koreader/fbink -y -6 -pm "Update successful :)" blah bluh && echo "OK" || echo "FAILED"
echo
date
echo Done
) >>/mnt/onboard/.adds/fbink.log 2>&1
```
and type in terminal emulator in koreader:
`(/mnt/onboard/.adds/misc/bg_strace.sh >/dev/null 2>&1 &)`
exit koreader, go to nickel, wait 2mn not seeing any message, exit Nickel, wait 30s in KSM, only then I see it displayed over KSM menu.
Or nickel is killall -STOP and killall -CONT when done ?
What's logged then:
```
Thu Jul  5 21:04:06 CEST 2018
Sleeping

Thu Jul  5 21:07:04 CEST 2018
Launching
[FBInk] Variable fb info: 1448x1072, 16bpp @ rotation: 0
[FBInk] Fontsize set to 32x32.
[FBInk] Line length: 45 cols, Page size: 33 rows.
[FBInk] Fixed fb info: smem_len 6782976, line_length 2944
[FBInk] Kernel's HZ value appears to be 100
Printing string 'Update successful :)' @ column 0, row -6 (inverted: false, flashing: false, centered: true, left padded: true, clear screen: false, font scaling: x0)
Printing string 'blah' @ column 0, row -5 (inverted: false, flashing: false, centered: true, left padded: true, clear screen: false, font scaling: x0)
Printing string 'bluh' @ column 0, row -4 (inverted: false, flashing: false, centered: true, left padded: true, clear screen: false, font scaling: x0)
OK

Thu Jul  5 21:07:04 CEST 2018
Done

```

**Comment by poire-z** [comment]
Behave the same on fw3.16 (on which I did some strace for you the same way, and that worked...).

But it happened after 5 tries that some bg fbink once displayed correctly over the onanimator animation.
Output:
```
Thu Jul  5 21:59:02 CEST 2018
Launching
[FBInk] Variable fb info: 1072x1448, 16bpp @ rotation: 3
[FBInk] Fontsize set to 32x32.
[FBInk] Line length: 33 cols, Page size: 45 rows.
[FBInk] Fixed fb info: smem_len 6782976, line_length 2176
[FBInk] Kernel's HZ value appears to be 100
Printing string 'Update successful :)' @ column 0, row -6 (inverted: false, flashing: false, centered: true, left padded: true, clear screen: false, font scaling: x0)
Printing string 'blah' @ column 0, row -5 (inverted: false, flashing: false, centered: true, left padded: true, clear screen: false, font scaling: x0)
Printing string 'bluh' @ column 0, row -4 (inverted: false, flashing: false, centered: true, left padded: true, clear screen: false, font scaling: x0)
OK
```

**Comment by NiLuJe** [comment]
@poire-z: Most likely I/O buffering quirks. Not quite sure how to deal with those in a shell script. I'd try ditching the subshell and doing command per command I/O redirections, but I'm not quite sure it'll change anything ;).

----

Unfortunately, I kinda need something from within nickel itself for real, because pickel does its own mess with the fb modes, as you've seen from the on-animator loop ;).

**Comment by poire-z** [comment]
Well, I can't make it happen... (not sure about I/O, my sleep 60 was taking 180s...)

Just in case it answers your question, in KSM9's `onstart/start_plato.sh`:
```sh
#! /bin/sh

# todo: set rotation value for the different device models
# currently I know only for sure that trilogy|kraken|alyssum|daylight need 3
case $PRODUCT in
  dragon|dahlia|snow ) platoRotate="1";;
  * ) platoRotate="3";;
esac

if [ "${platoRotate}" != "$(cat /sys/class/graphics/fb0/rotate)" ]; then
  echo "${platoRotate}" > /sys/class/graphics/fb0/rotate
  if [ "${platoRotate}" != "$(cat /sys/class/graphics/fb0/rotate)" ]; then
    echo "$((platoRotate ^ 2))" > /sys/class/graphics/fb0/rotate
  fi
fi
```

**Comment by NiLuJe** [comment]
No worries, I'm going to assume your final output was Nickel's, and hope for the best... :D.

I maaaaay be able to cobble something together in a test branch, we'll see ;).

I can't touch the hardware rotation myself, that'd obviously break everything else ;). I have to work it around inside FBInk without touching the HW... ;).

**Comment by NiLuJe** [comment]
Oookay, I may have the rotation thingy handled...

Can you try this binary? (Under KOReader, I mean ;p). It should (hopefully), print in black on white in the proper orientation (... as long as KOReader/KSM is in portrait mode. All bets are off if there's another layer of software rotation on top ^^).

[FBInk-v0.9.13-13-g1068f93.zip](https://github.com/koreader/koreader/files/2168603/FBInk-v0.9.13-13-g1068f93.zip)

**Comment by poire-z** [comment]
Working :) Correct orientation, no coloring, and the smiley stays cute:

<kbd>reader_2018-jul-06_073323</kbd>

```
/mnt/onboard/.adds # ./fbink -y -6 -pm "Update successful :)" blah bluh
[FBInk] Variable fb info: 1448x1072, 16bpp @ rotation: 0 (Upright, 0Â°)
[FBInk] Enabled Kobo 16bpp fb rotation device quirks
[FBInk] Fontsize set to 32x32.
[FBInk] Line length: 33 cols, Page size: 45 rows.
[FBInk] Fixed fb info: ID is "mxc_epdc_fb", length of fb mem: 6782976 bytes & line length: 2944 bytes
[FBInk] Kernel's HZ value appears to be 100
Printing string 'Update successful :)' @ column 0, row -6 (inverted: false, flashing: false, centered: true, left padded: true, clear screen: false, font scaling: x0)
Printing string 'blah' @ column 0, row -5 (inverted: false, flashing: false, centered: true, left padded: true, clear screen: false, font scaling: x0)
Printing string 'bluh' @ column 0, row -4 (inverted: false, flashing: false, centered: true, left padded: true, clear screen: false, font scaling: x0)
```
(Upright, 0Â° because my terminal is latin1 probably)

**Comment by NiLuJe** [comment]
Yay!

Many many thanks, that was basically the last thing preventing me from tagging a stable release, because I wasn't sure I could trust my H2O's 16bpp mode to work on this ;).

**Pull request merged**